### PR TITLE
Packing/unpacking lambdas

### PR DESCRIPTION
**Pull request by mehrdadn**
TODO: Latency performance benchmark

### CODE REVIEW

**Comment by mehrdadn** [comment]
TODO: Latency performance benchmark

File: install-dependencies.sh [hunk 1]
```diff
@@ -31,11 +31,11 @@ if [[ $platform == "linux" ]]; then
   # These commands must be kept in sync with the installation instructions.
   sudo apt-get update
   sudo apt-get install -y git cmake build-essential autoconf curl libtool python-dev python-numpy python-pip libboost-all-dev unzip graphviz
-  sudo pip install ipython typing funcsigs subprocess32 protobuf==3.0.0a2 colorama graphviz
+  sudo pip install ipython typing funcsigs subprocess32 protobuf==3.0.0a2 colorama graphviz cloudpickle
```
Comment (robertnishihara) [inline install-dependencies.sh]: please add cloudpickle to `doc/install-on-macosx` and `doc/install-on-ubuntu`

File: lib/python/ray/worker.py [hunk 2]
```diff
@@ -526,6 +533,7 @@ def remote(arg_types, return_types, worker=global_worker):
     return_types (List[type]): List of Python types of the return values.
   """
   def remote_decorator(func):
+    to_export = cloudpickle.dumps((func, arg_types, return_types)) if worker.mode == ray.SHELL_MODE else None
```
Comment (robertnishihara) [inline lib/python/ray/worker.py]: probably `if worker.mode == ray.SHELL_MODE or worker.mode == ray.SCRIPT_MODE`, basically whenever the worker is a driver

Comment (robertnishihara) [inline lib/python/ray/worker.py]: The `cloudpickle.dumps(...)` should probably be its own function (probably in a new file). Same with the `cloudpickle.loads(...)` above.

**Comment by robertnishihara** [comment]
We probably need to customize cloudpickle to handle remote functions a little differently. In `python scripts/shell.py`, the following does not work for me.
```python
In [1]: @ray.remote([int], [int])
   ...: def f(x):
   ...:     return x + 1
   ...:
In [2]: @ray.remote([int], [int])
   ...: def g(x):
   ...:     return ray.get(f(x)) + 1
   ...:
In [3]: ray.get(g(1))
1 task in progress
1 task in progress
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
scripts/shell.py in <module>()
----> 1 ray.get(g(1))

/Users/rkn/Workspace/ray/lib/python/ray/worker.pyc in get(objref, worker)
    400   value = worker.get_object(objref)
    401   if isinstance(value, RayFailedObject):
--> 402     raise Exception("The task that created this object reference failed with error message:\n{}".format(value.error_message))
    403   return value
    404 

Exception: The task that created this object reference failed with error message:
          Traceback (most recent call last):
            File "<ipython-input-2-f25a1c0dd15c>", line 3, in g
          NameError: global name 'f' is not defined
```

File: src/worker.cc [hunk 3]
```diff
@@ -17,16 +17,32 @@ inline WorkerServiceImpl::WorkerServiceImpl(const std::string& worker_address)
...
+Status WorkerServiceImpl::ImportFunction(ServerContext* context, const ImportFunctionRequest* request, ImportFunctionReply* reply) {
+  std::unique_ptr<WorkerMessage> message(new WorkerMessage());
+  message->function = request->function().implementation();
+  RAY_LOG(RAY_INFO, "importing function");
+  {
+    WorkerMessage* message_ptr = message.get();
+    RAY_CHECK(send_queue_.send(&message_ptr), "error sending over IPC");
+  }
+  message.release();
   return Status::OK;
```
Comment (robertnishihara) [inline src/worker.cc]: If the scheduler calls `ImportFunction` 1000 times in a row very quickly, what will happen?

File: src/worker.cc [hunk 4]
```diff
@@ -17,16 +17,32 @@ inline WorkerServiceImpl::WorkerServiceImpl(const std::string& worker_address)
...
+
+Status WorkerServiceImpl::ImportFunction(ServerContext* context, const ImportFunctionRequest* request, ImportFunctionReply* reply) {
+  std::unique_ptr<WorkerMessage> message(new WorkerMessage());
+  message->function = request->function().implementation();
+  RAY_LOG(RAY_INFO, "importing function");
+  {
+    WorkerMessage* message_ptr = message.get();
+    RAY_CHECK(send_queue_.send(&message_ptr), "error sending over IPC");
+  }
+  message.release();
```
Comment (robertnishihara) [inline src/worker.cc]: Is this deallocating the message before we know the message has been received on the other side?

File: install-dependencies.sh (hunk 1 above)
Comment (mehrdadn) [inline install-dependencies.sh]: Will do, but any way to have that one "include" a list or something so this doesn't have to be done in the future?

File: src/worker.cc (hunk 3 above)
Comment (mehrdadn) [inline src/worker.cc]: That's what should happen, they use the same message queue. Or are you saying it doesn't?

File: src/worker.cc (hunk 4 above)
Comment (mehrdadn) [inline src/worker.cc]: It's doing the opposite. `release` releases ownership of the memory, not the memory itself.

**Comment by mehrdadn** [comment]
Ah yeah, I'll try to handle that.

File: lib/python/ray/pickling.py [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import cloudpickle
+
+def dumps(func, arg_types, return_types):
+	return cloudpickle.dumps((func, arg_types, return_types))
```
Comment (robertnishihara) [inline lib/python/ray/pickling.py]: tabs => spaces

Comment (mehrdadn) [inline lib/python/ray/pickling.py]: Whoops, fixed! How do you even notice these?! :)

File: lib/python/ray/worker.py [hunk 6]
```diff
@@ -510,13 +512,24 @@ def process_task(task): # wrapping these lines in a function should cause the lo
...
-    task = ray.lib.wait_for_next_task(worker.handle)
-    if task is None:
-      # We use this as a mechanism to allow the scheduler to kill workers. When
-      # the scheduler wants to kill a worker, it gives the worker a null task,
-      # causing the worker program to exit the main loop here.
-      break
-    process_task(task)
+    (task, function) = ray.lib.wait_for_next_message(worker.handle)
+    try:
+      assert task is not None or function is not None
```
Comment (robertnishihara) [inline lib/python/ray/worker.py]: actually, this needs to be an "and" instead of an "or", right?

Comment (mehrdadn) [inline lib/python/ray/worker.py]: Wait, why though? It should work just fine if neither is `None`.

**Pull request merged**